### PR TITLE
Support arbitrary number of EDPs and enable cumulative discrepancies

### PR DESCRIPTION
**Pull request by uakyol**
1. Dynamically creates cover and subset relationships to enable reports with arbitrary number of EDPs
2. Enables given EDP to violate cumulative rules

### CODE REVIEW

File: dp_consistency/tests/report/test_report.py [hunk 1]
```diff
@@ -3,114 +3,176 @@
 from src.noiseninja.noised_measurements import Measurement
 from src.report.report import Report, MetricReport
 
-EXPECTED_PRECISION = 4
+EXPECTED_PRECISION = 3
```
Comment (uakyol) [inline dp_consistency/tests/report/test_report.py:6]: I had to change this, it was failing with 4 I think this is incidental. ptal.

File: dp_consistency/tests/report/test_report.py [hunk 2]
```diff
@@ -3,114 +3,176 @@
...
+                        frozenset({EDP_ONE, EDP_TWO}): [Measurement(51.5, 1)],
+                        frozenset({EDP_ONE}): [Measurement(50.5, 1)],
+                    }
+                ),
                 mrc: MetricReport(
-                    total_campaign_reach_time_series=[Measurement(51, 1)],
-                    reach_time_series_by_edp=[
-                        [Measurement(51, 1)]])},
+                    reach_time_series_by_edp_combination={
+                        frozenset({EDP_ONE, EDP_TWO}): [Measurement(51.5, 1)],
```
Comment (uakyol) [inline dp_consistency/tests/report/test_report.py:134]: Can you please double check the numbers in this test. I had to change them, I think with this new approach optimizer took a different route but the results are still valid.

File: dp_consistency/src/noiseninja/solver.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 
 from src.noiseninja.noised_measurements import SetMeasurementsSpec
 
-SOLVER = "clarabel"
+SOLVER = "highs"
```
Comment (uakyol) [inline dp_consistency/src/noiseninja/solver.py:9]: Without this change, this code is broken at head :( 
Lets keep this.

Comment (rmirisola) [reply dp_consistency/src/noiseninja/solver.py:9]: Could you make this a flag of the reporting tool?

Doesn't have to be on this PR.

File: dp_consistency/tests/report/test_report.py (hunk 2 above)
Comment (rmirisola) [reply dp_consistency/tests/report/test_report.py:285]: Old test case has higher squared error I think, so this is good.

File: dp_consistency/src/report/report.py [hunk 4]
```diff
@@ -4,112 +4,176 @@
...
+        for comb1, comb2 in combinations(edp_combinations, 2):
+            if comb1.issubset(comb2):
+                subset_relationships.append((comb2, comb1))
+            elif comb2.issubset(comb1):
+                subset_relationships.append((comb1, comb2))
+        return subset_relationships
+
+    def get_cover_relationships(self):
+        """Returns covers as defined here: # https://en.wikipedia.org/wiki/Cover_(topology).
+        For each set (s_i) in the list, enumerate combinations of all sets excluding this one.
```
Comment (rmirisola) [inline dp_consistency/src/report/report.py:80]: you can probably get a quick performance win by only including subsets of S in the list of subsets you look for combinations of.

You can probably make it even better by introducing some memoization and noting that if A and B cover C the so must all covers(A) X covers(B)

Comment (rmirisola) [reply dp_consistency/src/report/report.py:80]: you may want to add a unit test for this given this is the most complex logic here.

File: dp_consistency/tests/tools/test_correctoriginreport.py [hunk 5]
```diff
@@ -10,8 +10,7 @@
 MRC_FILTER = "MRC"
 
 
-class TestReport(TestCase):
-
+class TestOriginSheetReport(TestCase):
```
Comment (rmirisola) [inline dp_consistency/tests/tools/test_correctoriginreport.py:13]: did you include an updated test for N EDPs?

File: dp_consistency/src/report/report.py (hunk 4 above)
Comment (ple13) [reply dp_consistency/src/report/report.py:80]: +1. The current algorithm may have O(n2^{n-1}) computation time, which will not scale well when there are more EDPs.
typo: union -> union

File: dp_consistency/src/report/report.py [hunk 6]
```diff
@@ -4,112 +4,176 @@
...
+        """
+
+        def generate_all_length_combinations(data):
+            return [
+                comb for r in range(1, len(data) + 1) for comb in combinations(data, r)
+            ]
+
+        cover_relationships = []
+        edp_combinations = list(self.__reach_time_series_by_edp_combination)
+        for i in range(len(edp_combinations)):
```
Comment (ple13) [inline dp_consistency/src/report/report.py:92]: Please let me know if my understanding is correct: Let the input be { (a, b, c), (a, b), (a, c), (b, c) }, then the output will be { [(a, b), (b, c), (c, a)] covers (a, b, c) }? Should we have tighter relationships such as: { [(a, b), (b, c)] covers (a, b, c); etc. }

File: dp_consistency/tests/tools/test_correctoriginreport.py (hunk 5 above)
Comment (uakyol) [reply dp_consistency/tests/tools/test_correctoriginreport.py:13]: Not doing this per offline convo, will tackle this for the report proto

File: dp_consistency/src/report/report.py (hunk 6 above)
Comment (uakyol) [reply dp_consistency/src/report/report.py:92]: Output will be both [(a, b), (b, c), (c, a)] covers (a, b, c)  
AND
[(a, b), (b, c)] covers (a, b, c) 

I detailed this in a test case.

File: dp_consistency/src/report/report.py [hunk 7]
```diff
@@ -4,112 +4,179 @@
...
+            possible_covered = edp_combinations[i]
+            other_sets = edp_combinations[:i] + edp_combinations[i + 1 :]
+            all_subsets_of_possible_covered = [other_set for other_set in other_sets if other_set.issubset(possible_covered)]
+            possible_covers = generate_all_length_combinations(all_subsets_of_possible_covered)
+            for possible_cover in possible_covers:
+                union_of_possible_cover = reduce(
+                    lambda x, y: x.union(y), possible_cover
+                )
+                if union_of_possible_cover == possible_covered:
+                    # print("possible_coveredpossible_coveredpossible_coveredpossible_covered")
```
Comment (ple13) [inline dp_consistency/src/report/report.py:102]: nit: remove the commented out `print`.

**Review by ple13 — APPROVED** [review]

**Review by rmirisola — APPROVED** [review]

**Pull request merged**